lib/bamboo/view.ex

```diff
@@ -0,0 +1,201 @@
+
+      #{available_templates(exception.available)}
+      """
+    end
+
+    defp available_templates([]), do: "No templates were compiled for this module."
+
+    defp available_templates(available) do
+      "The following templates were compiled:\n\n" <>
+        Enum.map_join(available, "\n", &"* #{&1}") <>
+        "\n"
+    end
+  end
+
+  defmacro __using__(path: path) do
+    quote do
+      @templates_path unquote(path)
+      @before_compile unquote(__MODULE__)
+    end
+  end
+
+  defmacro __using__(opts) do
+    raise ArgumentError, """
+    expected Bamboo.View to have a path set, instead got: #{inspect(opts)}.
+
+    Please set a path where this view's template are defined e.g. use Bamboo.View, path: "/lib/my_app/email/templates/account"
+    """
+  end
+
+  defmacro __before_compile__(env) do
+    templates_path = Module.get_attribute(env.module, :templates_path)
+
+    results =
+      templates_path
+      |> templates_to_compile()
+      |> Enum.map(&template_and_name/1)
+      |> Enum.map(fn {template, name} ->
+        {name, compile(template, name)}
+      end)
+
+    template_names = Enum.map(results, &elem(&1, 0))
+    render_functions = Enum.map(results, &elem(&1, 1))
+
+    quote do
+      unquote(render_functions)
+
+      def render_template(template, _assigns) do
+        raise UndefinedTemplateError,
+          module: __MODULE__,
+          template: template,
+          available: unquote(template_names)
+      end
+    end
+  end
+
+  @doc false
+  def render_email(view, email, template, assigns) do
+    email
+    |> put_defaults(view)
+    |> merge_assigns(assigns)
+    |> render(template)
+  end
+
+  defp put_defaults(%{private: private} = email, view_module) do
+    private =
+      private
+      |> Map.put_new(:html_layout, false)
+      |> Map.put_new(:text_layout, false)
+      |> Map.put_new(:view_module, view_module)
+
+    %{email | private: private}
+  end
+
+  defp merge_assigns(%{assigns: email_assigns} = email, assigns) do
+    assigns = email_assigns |> Map.merge(Enum.into(assigns, %{}))
+    email |> Map.put(:assigns, assigns)
+  end
+
+  defp render(email, template) when is_atom(template) do
+    render_html_and_text_emails(email, template)
+  end
+
+  defp render(email, template) do
+    render_text_or_html_email(email, template)
+  end
+
+  defp render_html_and_text_emails(email, template) do
+    view_template = Atom.to_string(template)
+
+    email
+    |> Map.put(:html_body, render_html(email, view_template <> ".html"))
+    |> Map.put(:text_body, render_text(email, view_template <> ".text"))
+  end
+
+  defp render_text_or_html_email(email, template) do
+    cond do
+      String.ends_with?(template, ".html") ->
+        email |> Map.put(:html_body, render_html(email, template))
+
+      String.ends_with?(template, ".text") ->
+        email |> Map.put(:text_body, render_text(email, template))
+
+      true ->
+        raise ArgumentError, """
```

Comment:
Is it worth defining a exception here? `InvalidTemplateTypeError`?

Reply:
That's a good question. I think the error is good enough as is. It's an argument error because you passed the wrong argument to `render/2` -> since the template doesn't exist. But I could be convinced otherwise. Not sure I have a strong opinion here.